### DISCUSSION TITLE
SmtpError 400 Bad Request

### DISCUSSION THREAD

**Discussion by vwlau** [dani-garcia/vaultwarden#1674] [Q&A] [answered]
I've looked at half a dozen other posts about smtp and none of the solutions in them have worked. I'm running vaultwarden as a docker container on a remote server behind an nginx reverse proxy. Support string below:

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.21.0
* Web-vault version: v2.19.0d
* Running within Docker: true
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.33.0
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": false,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_ip_header_enabled": true,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_max_conns": 10,
  "database_url": "****/**.*******",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://**.****.**",
  "domain_origin": "*****://**.****.**",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "enable_db_wal": true,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "invitation_org_name": "Vaultwarden",
  "invitations_allowed": false,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": null,
  "log_level": "Info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "org_attachment_limit": null,
  "org_creation_users": "",
  "password_iterations": 100000,
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sends_folder": "data/sends",
  "show_password_hint": true,
  "signups_allowed": true,
  "signups_domains_whitelist": "",
  "signups_verify": false,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_explicit_tls": false,
  "smtp_from": "************@*****.***",
  "smtp_from_name": "Vaultwarden",
  "smtp_host": "****.*****.***",
  "smtp_password": "***",
  "smtp_port": 587,
  "smtp_ssl": true,
  "smtp_timeout": 15,
  "smtp_username": "************@*****.***",
  "templates_folder": "data/templates",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_syslog": false,
  "user_attachment_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": false,
  "websocket_port": 3012,
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```

I get this when trying to send a test email.

```
[CAUSE] lettre::transport::smtp::Error {
     kind: Connection,
     source: "Could not connect",
 }
 [2021-05-07 18:09:07.659][response][INFO] POST /admin/test/smtp (test_smtp) => 400 Bad Request
```

I've tried gmail and sendgrid. I remembered to save every time I made changes to the config page. I've enabled "less secure apps" and I tried [this solution](https://github.com/dani-garcia/vaultwarden/discussions/1360#discussioncomment-342031) for gmail smtp but it did not work.

I tried `telnet smtp.gmail.com 587` and get this response.
```
Trying 173.194.204.108...
Trying 2607:f8b0:400d:c07::6c...
telnet: Unable to connect to remote host: Network is unreachable
```

Does that mean my ports are being blocked by the ISP? 

Any help would be appreciated.

**Comment by BlackDex** [comment]
That could be. Have you also tried setting tls to true and setting the port to 465?

**Reply by vwlau** [reply]
Yeah, I've tried every combination of ssl, tls, and port numbers 587 & 465.

**Reply by BlackDex** [reply]
If that and telnet and openssl s_client all not work it's probably some outgoing firewall

**Reply by MilesTEG1** [reply]
Hello,
In my docker-compose file, I set up this two environment variables :
```yaml
      # Pour l'envoi d'emails
      - SMTP_HOST=smtp.gmail.com
      - SMTP_FROM=XXxxXX
      - SMTP_FROM_NAME=BlaBla
      - SMTP_PORT=587
      - SMTP_SSL=true
      - SMTP_USERNAME=XXxxXX
      - SMTP_PASSWORD=XXxxXX
```
Vaultwarden can send emails without problems.
As I use 2FA, I had to generate an app pasword for Vaultwarden.
Did you do the same thing ?

**Reply by vwlau** [reply]
> If that and telnet and openssl s_client all not work it's probably some outgoing firewall

Not sure what I should be trying with openssl s_client but I tried this command `openssl s_client -connect smtp.google.com:587` and got this response:

```
140557678978176:error:2008A067:BIO routines:BIO_connect:connect error:../crypto/bio/b_sock2.c:111:
140557678978176:error:0200206E:system library:connect:Connection timed out:../crypto/bio/b_sock2.c:110:
140557678978176:error:2008A067:BIO routines:BIO_connect:connect error:../crypto/bio/b_sock2.c:111:
140557678978176:error:0200206E:system library:connect:Connection timed out:../crypto/bio/b_sock2.c:110:
140557678978176:error:2008A067:BIO routines:BIO_connect:connect error:../crypto/bio/b_sock2.c:111:
140557678978176:error:0200206E:system library:connect:Connection timed out:../crypto/bio/b_sock2.c:110:
140557678978176:error:2008A067:BIO routines:BIO_connect:connect error:../crypto/bio/b_sock2.c:111:
140557678978176:error:02002065:system library:connect:Network is unreachable:../crypto/bio/b_sock2.c:110:
140557678978176:error:2008A067:BIO routines:BIO_connect:connect error:../crypto/bio/b_sock2.c:111:
140557678978176:error:02002065:system library:connect:Network is unreachable:../crypto/bio/b_sock2.c:110:
140557678978176:error:2008A067:BIO routines:BIO_connect:connect error:../crypto/bio/b_sock2.c:111:
140557678978176:error:02002065:system library:connect:Network is unreachable:../crypto/bio/b_sock2.c:110:
140557678978176:error:2008A067:BIO routines:BIO_connect:connect error:../crypto/bio/b_sock2.c:111:
140557678978176:error:02002065:system library:connect:Network is unreachable:../crypto/bio/b_sock2.c:110:
140557678978176:error:2008A067:BIO routines:BIO_connect:connect error:../crypto/bio/b_sock2.c:111:
connect:errno=101
```

I have ufw set to allow 587

**Reply by vwlau** [reply]
> Hello,
> In my docker-compose file, I set up this two environment variables :
> 
> ```yaml
>       # Pour l'envoi d'emails
>       - SMTP_HOST=smtp.gmail.com
>       - SMTP_FROM=XXxxXX
>       - SMTP_FROM_NAME=BlaBla
>       - SMTP_PORT=587
>       - SMTP_SSL=true
>       - SMTP_USERNAME=XXxxXX
>       - SMTP_PASSWORD=XXxxXX
> ```
> 
> Vaultwarden can send emails without problems.
> As I use 2FA, I had to generate an app pasword for Vaultwarden.
> Did you do the same thing ?

I initially did not have 2FA enabled and had "less secure apps" on. I then tried with 2FA on and created an app password and that didnt work either.

**Reply by BlackDex** [reply]
For the openssl command you probably also need `-starttls smtp` as addition with port 587.

You could try to disable ufw for a short while and see if that solves it of course.

**Reply by vwlau** [reply]
I did some googling and I believe my vps provider blocks smtp ports but I can ask to be whitelisted. I will report back after I find out for sure.

**Reply by vwlau** [reply]
> For the openssl command you probably also need `-starttls smtp` as addition with port 587.
> 
> You could try to disable ufw for a short while and see if that solves it of course.

`openssl s_client -connect smtp.gmail.com:587 -starttls smtp`

Gave me the same error messages as above.

**Comment by vwlau** [comment, marked as answer]
It turns out my VPS provider had blocked the SMTP ports. I asked to have my VPS whitelisted and they opened them for me. Email now sends out with the configuration I provided in the original question.
